core/consensus/common/src/import_queue.rs

```diff
@@ -698,9 +615,7 @@ mod tests {
 		let _ = importer_sender.send(BlockImportMsg::Start(Box::new(link.clone())));
 
 		// Ensure the importer handles Start before any result messages.
-		let (ack_sender, ack_port) = channel::unbounded();
-		let _ = importer_sender.send(BlockImportMsg::Status(ack_sender));
-		let _ = ack_port.recv();
+		thread::sleep(Duration::from_millis(1));
```

Comment:
Why do we need to sleep here? I assume you want to ensure that something is setup in another thread?

In certain situations, this will break the test. Sleeping in a test is never a good idea.

Reply:
We need the `select` of the `BlockImporter` to handle the `Start` message first. If we don't sleep here, it might handle the `BlockImportWorkerMsg::Imported` first (since the `select` randomly chooses either channels), and this will result in an intermittent panic for this test, since no worker is actually running. 

In production this cannot happen because a `BlockImportWorkerMsg::Imported` will only be received if the importer first sends an `BlockImportWorkerMsg::ImportBlocks` message, and the worker will actually be running.

Reply:
Can we check this in a different way? In some weird situations this may lead to test breakage.

Reply:
I can't think of a different way to check this at this stage, let me know if you have ideas. 

The issue here is that we're sending two messages on two different channels, which the other thread selects over. In real code, there is no problem, but in this test we need to ensure that one message is selected before the other. The problem arises when the select wakes up and needs to randomly choose one message or the other. If we sleep for even 1 ms after sending the first message, the problem goes away since the select will wake up to only one message in one channel.

Reply:
Multithreaded code does not work this way and is not predictable. We can go this way, but maybe need to revisit this when the test breaks once.

Reply:
@gterzian: If you need to introduce ordering `BlockImportMsg::Start` could be extended to include a (oneshot) Sender to report back when it has finished starting. You would then wait on the receiving side and only then proceed sending messages over the other channel.

Reply:
@twittner It's a good idea(as you see from the diff something similar was previously done using the now removed `Status` message). 

I would do it if it were also needed outside of this test, and otherwise I'd rather not make further changes to the non-test code to support this specific test case. Also, if we don't actually need to block on the `Start` message outside of tests, I don't think we should do it just to support this one test.  In this specific case the `sleep` statement will do just fine because:

1. Outside of this specific tests, this problem cannot happen(we're simulating receiving import results without having an actual import worker running, the ordering we're after happens naturally if the worker is actually running).
2. The sleep statement is the most direct solution to this problem. It introduces the deterministic ordering we're looking for in this situation only, and I haven't found opportunities for intermittent failures yet.

Reply:
Sleep does not introduce deterministic ordering in multithreaded situations, never! You don't know how the OS is scheduling your threads and on high load everything could happen. That is something that could easily evolve to a flaky test.

If there was support for doing this without sleep before, please restore this behavior.

Reply:
Perhaps not quite deterministic, but good enough. It's unlikely that the scheduling of threads would make this test fail. 

The `send` on the channel will ensure the other thread "wakes-up" via `unpark`. 

So that the other thread will not be scheduled while the main thread sleeps, is unlikely. That the main thread were to be re-scheduled before the other thread, even less so. 

We could simply increase the `sleep` to 10ms instead of 1 to get 99% certainty the test never would fail. 

"under load" anything could happen indeed, and are we writing this to perform under load? it's a unit-test. 

On the other hand, removing things like the `Status` message that was previously used instead of `sleep`, is something that will affect how the production code performs under load.

Reply:
Ty